README.md
# auto_validating_text_field

A Flutter package which validates text field dynamically as soon as value changes. You can give list of validators as shown below in example to dynamically validate according to the validators. 

## Usage
To use this package, add              

  **auto_validating_text_field** :  

as a dependency in your pubspec.yaml file.

## Example
   
    AutoValidatingTextFormField(
      decoration: InputDecoration(
      hintText: "Full Name",
      ),
      focusNode: _fullNameFocusNode,
      controller: _fullNameController,
      validators: [
        Validators.required(),
        Validators.pattern("^[a-zA-Z ]*\$",
          errorText: "Full Name cannot contains special characters."),
      ]
    );

Please see the example app of this package for a full example in 
https://github.com/rmsh03dhj/auto_validating_text_field/tree/master/example/lib.


CHANGELOG.md
## [0.0.1] - 2020-05-23
   Initial Open Source release
   
## [0.0.2] - 2020-05-23
   Adding example
   
 ## [0.1.0] - 2020-05-31
   First stable release
   
  ## [0.1.1] - 2020-05-31
   Updating rx dart package
   
  ## [0.1.2] - 2020-06-14
   Updating logic to consider the empty value once user deletes the content
   
  ## [0.1.3] - 2020-06-21
   Supporting older version of rxDart as well.
   
  ## [0.1.4] - 2020-07-12
   Updating rx dart package and adding logic to determine whether to auto validate or not initially depending upon textfield value. If its empty, it doesn't validate else it will validate.



example/lib/main.dart
import 'package:flutter/material.dart';

import 'package:auto_validating_text_field/auto_validating_text_field.dart';
import 'Validators.dart';
import 'sign_up_dto.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: SignUpScreen(),
    );
  }
}

class SignUpScreen extends StatefulWidget {
  @override
  _SignUpScreen createState() => _SignUpScreen();
}

class _SignUpScreen extends State<SignUpScreen> {
  TextEditingController _fullNameController = TextEditingController();
  TextEditingController _phoneNoController = TextEditingController();
  TextEditingController _emailAddressController = TextEditingController();
  TextEditingController _passwordController = TextEditingController();
  TextEditingController _confirmPasswordController = TextEditingController();

  FocusNode _fullNameFocusNode = FocusNode();
  FocusNode _phoneNumberFocusNode = FocusNode();
  FocusNode _emailAddressFocusNode = FocusNode();
  FocusNode _passwordFocusNode = FocusNode();
  FocusNode _confirmPasswordFocusNode = FocusNode();

  final _formKey = GlobalKey<FormState>();
  final GlobalKey<ScaffoldState> _scaffoldKey = new GlobalKey<ScaffoldState>();

  SignUpDto signUpDetails;
  @override
  void initState() {
    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      key: _scaffoldKey,
      body: SingleChildScrollView(
        child: Form(
          key: _formKey,
          child: Column(
            children: <Widget>[
              Container(
                height: 30,
              ),
              Padding(
                padding: const EdgeInsets.all(8.0),
                child: Center(
                  child: Text(
                    "Sign Up",
                    style: TextStyle(
                      fontSize: 30,
                    ),
                  ),
                ),
              ),
              SizedBox(height: 50),
              Container(
                width: 250,
                child: AutoValidatingTextFormField(
                  decoration: InputDecoration(
                    hintText: "Full Name",
                  ),
                  focusNode: _fullNameFocusNode,
                  controller: _fullNameController,
                  keyboardType: TextInputType.text,
                  textCapitalization: TextCapitalization.words,
                  validators: [
                    Validators.required(),
                    Validators.pattern("^[a-zA-Z ]*\$",
                        errorText:
                        "Full Name cannot contains special characters."),
                  ],
                  onFieldSubmitted: (_) {
                    fieldFocusChange(
                        context, _fullNameFocusNode, _phoneNumberFocusNode);
                  },
                ),
              ),
              SizedBox(height: 20),
              Container(
                width: 250,
                child: AutoValidatingTextFormField(
                  decoration: InputDecoration(
                    hintText: "Phone Number",
                  ),
                  controller: _phoneNoController,
                  focusNode: _phoneNumberFocusNode,
                  keyboardType: TextInputType.number,
                  validators: [
                    Validators.required(),
                    Validators.validLength(5,
                        errorText: "Phone number should be of length 5"),
                  ],
                  onFieldSubmitted: (_) {
                    fieldFocusChange(
                        context, _phoneNumberFocusNode, _emailAddressFocusNode);
                  },
                ),
              ),
              SizedBox(height: 20),
              Container(
                width: 250,
                child: AutoValidatingTextFormField(
                  decoration: InputDecoration(
                    hintText: "Email Addres",
                  ),
                  focusNode: _emailAddressFocusNode,
                  controller: _emailAddressController,
                  keyboardType: TextInputType.emailAddress,
                  validators: [
                    Validators.required(),
                    Validators.pattern(
                        "^[a-zA-Z0-9.!#\$%&'*+/=?^_`{|}~-]+@[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,253}[a-zA-Z0-9])?(?:\.[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,253}[a-zA-Z0-9])?)*\$",
                        errorText: "Invalid email address."),
                  ],
                  onFieldSubmitted: (_) {
                    fieldFocusChange(
                        context, _emailAddressFocusNode, _passwordFocusNode);
                  },
                ),
              ),
              SizedBox(height: 20),
              Container(
                width: 250,
                child: AutoValidatingTextFormField(
                  decoration: InputDecoration(
                    hintText: "Password",
                  ),
                  obscureText: true,
                  focusNode: _passwordFocusNode,
                  controller: _passwordController,
                  keyboardType: TextInputType.text,
                  validators: [
                    Validators.required(),
                  ],
                  onFieldSubmitted: (_) {
                    fieldFocusChange(
                        context, _passwordFocusNode, _confirmPasswordFocusNode);
                  },
                ),
              ),
              SizedBox(height: 20),
              Container(
                width: 250,
                child: AutoValidatingTextFormField(
                  decoration: InputDecoration(
                    hintText: "Confirm Password",
                  ),
                  keyboardType: TextInputType.text,
                  focusNode: _confirmPasswordFocusNode,
                  controller: _confirmPasswordController,
                  obscureText: true,
                  validators: [
                    Validators.required(),
                    Validators.matches(
                      _passwordController,
                    )
                  ],
                ),
              ),
              SizedBox(height: 40),
              RaisedButton(
                child: Text(
                  "Sign Up",
                ),
                onPressed: () {
                  if (_formKey.currentState.validate()) {
                    _scaffoldKey.currentState.showSnackBar(
                      SnackBar(
                        content: Text(
                          "Signing up",
                        ),
                        duration: const Duration(seconds: 1),
                      ),
                    );
                  }
                },
              ),
            ],
          ),
        ),
      ),
    );
  }

  @override
  void dispose() {
    super.dispose();
  }
}

void fieldFocusChange(
    BuildContext context, FocusNode currentFocus, FocusNode nextFocus) {
  currentFocus.unfocus();
  FocusScope.of(context).requestFocus(nextFocus);
}




example/lib/Validators.dart
import 'package:flutter/material.dart';

class Validators {
  static FormFieldValidator required(
      {String errorText = "This field cannot be empty."}) {
    return (value) {
      if (value == null ||
          ((value is Iterable || value is String || value is Map) &&
              value.length == 0)) {
        return errorText;
      }
      return null;
    };
  }

  static FormFieldValidator validLength(int length,
      {String errorText = "Invalid phone number"}) {
    return (value) {
      if (value != null && value.isNotEmpty) {
        if (value.length != length) {
          return errorText;
        }
      }
      return null;
    };
  }

  static FormFieldValidator pattern(
    Pattern pattern, {
    String errorText = "Value does not match pattern.",
  }) {
    return (value) {
      if (value != null && value.isNotEmpty) {
        if (!RegExp(pattern).hasMatch(value)) return errorText;
      }
      return null;
    };
  }

  static FormFieldValidator matches(TextEditingController controller, {
    String errorText = "Confirm Password doesn't match.",
  }) {
    return (value) {
      if (value != null && value.isNotEmpty) {
        if (controller.text.compareTo(value.toString()) != 0) return errorText;
      }
      return null;
    };
  }
}


lib/auto_validating_text_field.dart
library auto_validating_text_field;

import 'dart:core';

import 'package:flutter/material.dart';
import 'package:flutter/services.dart';
import 'package:rxdart/rxdart.dart';

class AutoValidatingTextFormField extends StatefulWidget {
  final TextEditingController controller;
  final List<FormFieldValidator> validators;
  final Key key;
  final FocusNode focusNode;

  final String initialValue;
  final InputDecoration decoration;
  final TextInputType keyboardType;
  final TextCapitalization textCapitalization;
  final TextInputAction textInputAction;
  final TextStyle style;
  final StrutStyle strutStyle;
  final TextDirection textDirection;
  final TextAlign textAlign;
  final TextAlignVertical textAlignVertical;
  final autoFocus;
  final readOnly;
  final ToolbarOptions toolbarOptions;
  final showCursor;
  final obscureText;
  final autoCorrect;
  final SmartDashesType smartDashesType;
  final SmartQuotesType smartQuotesType;
  final enableSuggestions;
  final autoValidate;
  final maxLengthEnforced;
  final int maxLines;
  final int minLines;
  final expands;
  final int maxLength;
  final ValueChanged<String> onChanged;
  final GestureTapCallback onTap;
  final VoidCallback onEditingComplete;
  final ValueChanged<String> onFieldSubmitted;
  final FormFieldSetter<String> onSaved;
  final FormFieldValidator<String> validator;
  final List<TextInputFormatter> inputFormatters;
  final enabled;
  final double cursorWidth;
  final Radius cursorRadius;
  final Color cursorColor;
  final Brightness keyboardAppearance;
  final EdgeInsets scrollPadding;
  final enableInteractiveSelection;
  final InputCounterWidgetBuilder buildCounter;
  final ScrollPhysics scrollPhysics;

  AutoValidatingTextFormField({
    this.controller,
    this.validators = const [],
    this.key,
    this.focusNode,
    this.initialValue,
    this.decoration = const InputDecoration(),
    this.keyboardType,
    this.textCapitalization = TextCapitalization.none,
    this.textInputAction,
    this.style,
    this.strutStyle,
    this.textDirection,
    this.textAlign = TextAlign.start,
    this.textAlignVertical,
    this.autoFocus = false,
    this.readOnly = false,
    this.toolbarOptions,
    this.showCursor,
    this.obscureText = false,
    this.autoCorrect = true,
    this.smartDashesType,
    this.smartQuotesType,
    this.enableSuggestions = true,
    this.autoValidate = false,
    this.maxLengthEnforced = true,
    this.expands = false,
    this.maxLength,
    this.onChanged,
    this.onTap,
    this.onEditingComplete,
    this.onFieldSubmitted,
    this.onSaved,
    this.validator,
    this.inputFormatters,
    this.enabled = true,
    this.cursorWidth = 2.0,
    this.cursorRadius,
    this.cursorColor,
    this.keyboardAppearance,
    this.scrollPadding = const EdgeInsets.all(20.0),
    this.enableInteractiveSelection = true,
    this.buildCounter,
    this.scrollPhysics,
    this.maxLines = 1,
    this.minLines,
  });

  @override
  _AutoValidatingTextFormFieldState createState() => _AutoValidatingTextFormFieldState();
}

class _AutoValidatingTextFormFieldState extends State<AutoValidatingTextFormField> {
  final GlobalKey<FormFieldState> _fieldKey = GlobalKey<FormFieldState>();
  bool _autoValidate = false;
  final _validationRequired = BehaviorSubject<bool>();

  @override
  void initState() {
    super.initState();

    if (widget.controller.text.isNotEmpty) {
      setState(() {
        _autoValidate = true;
      });
    } else {
      _validationRequired.add(false);
    }

    ///starts validating as soon as value changes
    widget.controller.addListener(() {
      if (widget.controller.text.isNotEmpty) {
        _validationRequired.add(true);

        if (_validationRequired.value && widget.focusNode.hasFocus) {
          if (!_fieldKey.currentState.validate()) {
            setState(() {
              _autoValidate = true;
            });
          }
        }
      } else {
        if (_validationRequired.value) {
          setState(() {
            _autoValidate = true;
          });
        } else {
          setState(() {
            _autoValidate = false;
          });
        }
      }
    });
  }

  @override
  Widget build(BuildContext context) {
    return TextFormField(
      key: _fieldKey,
      focusNode: widget.focusNode,
      autovalidate: _autoValidate,
      controller: widget.controller,
      validator: (val) {
        for (int i = 0; i < widget.validators.length; i++) {
          if (widget.validators[i](val) != null) return widget.validators[i](val);
        }
        return null;
      },
      decoration: widget.decoration,
      keyboardType: widget.keyboardType,
      textInputAction: widget.textInputAction,
      style: widget.style,
      strutStyle: widget.strutStyle,
      textAlign: widget.textAlign,
      textAlignVertical: widget.textAlignVertical,
      textDirection: widget.textDirection,
      textCapitalization: widget.textCapitalization,
      autofocus: widget.autoFocus,
      toolbarOptions: widget.toolbarOptions,
      readOnly: widget.readOnly,
      showCursor: widget.showCursor,
      obscureText: widget.obscureText,
      autocorrect: widget.autoCorrect,
      smartDashesType: widget.smartDashesType,
      smartQuotesType: widget.smartQuotesType,
      enableSuggestions: widget.enableSuggestions,
      maxLengthEnforced: widget.maxLengthEnforced,
      maxLines: widget.maxLines,
      minLines: widget.minLines,
      expands: widget.expands,
      maxLength: widget.maxLength,
      onChanged: widget.onChanged,
      onTap: widget.onTap,
      onEditingComplete: widget.onEditingComplete,
      onFieldSubmitted: widget.onFieldSubmitted,
      inputFormatters: widget.inputFormatters,
      enabled: widget.enabled,
      cursorWidth: widget.cursorWidth,
      cursorRadius: widget.cursorRadius,
      cursorColor: widget.cursorColor,
      scrollPadding: widget.scrollPadding,
      scrollPhysics: widget.scrollPhysics,
      keyboardAppearance: widget.keyboardAppearance,
      enableInteractiveSelection: widget.enableInteractiveSelection,
      buildCounter: widget.buildCounter,
    );
  }

  @override
  void dispose() {
    super.dispose();
    _validationRequired.close();
  }
}


example/lib/sign_up_screen.dart
import 'package:auto_validating_text_field/auto_validating_text_field.dart';
import 'package:flutter/material.dart';

import 'Validators.dart';
import 'sign_up_dto.dart';

class SignUpScreen extends StatefulWidget {
  @override
  _SignUpScreen createState() => _SignUpScreen();
}

class _SignUpScreen extends State<SignUpScreen> {
  TextEditingController _fullNameController = TextEditingController();
  TextEditingController _phoneNoController = TextEditingController();
  TextEditingController _emailAddressController = TextEditingController();
  TextEditingController _passwordController = TextEditingController();
  TextEditingController _confirmPasswordController = TextEditingController();

  FocusNode _fullNameFocusNode = FocusNode();
  FocusNode _phoneNumberFocusNode = FocusNode();
  FocusNode _emailAddressFocusNode = FocusNode();
  FocusNode _passwordFocusNode = FocusNode();
  FocusNode _confirmPasswordFocusNode = FocusNode();

  final _formKey = GlobalKey<FormState>();
  final GlobalKey<ScaffoldState> _scaffoldKey = new GlobalKey<ScaffoldState>();

  SignUpDto signUpDetails;
  @override
  void initState() {
    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      key: _scaffoldKey,
      body: SingleChildScrollView(
        child: Form(
          key: _formKey,
          child: Column(
            children: <Widget>[
              Container(
                height: 30,
              ),
              Padding(
                padding: const EdgeInsets.all(8.0),
                child: Center(
                  child: Text(
                    "Sign Up",
                    style: TextStyle(
                      fontSize: 30,
                    ),
                  ),
                ),
              ),
              SizedBox(height: 50),
              Container(
                width: 250,
                child: AutoValidatingTextFormField(
                  decoration: InputDecoration(
                    hintText: "Full Name",
                  ),
                  focusNode: _fullNameFocusNode,
                  controller: _fullNameController,
                  keyboardType: TextInputType.text,
                  textCapitalization: TextCapitalization.words,
                  validators: [
                    Validators.required(),
                    Validators.pattern("^[a-zA-Z ]*\$",
                        errorText:
                            "Full Name cannot contains special characters."),
                  ],
                  onFieldSubmitted: (_) {
                    fieldFocusChange(
                        context, _fullNameFocusNode, _phoneNumberFocusNode);
                  },
                  onSaved: (value) {
                    signUpDetails.fullName = value;
                  },
                ),
              ),
              SizedBox(height: 20),
              Container(
                width: 250,
                child: AutoValidatingTextFormField(
                  decoration: InputDecoration(
                    hintText: "Phone Number",
                  ),
                  controller: _phoneNoController,
                  focusNode: _phoneNumberFocusNode,
                  keyboardType: TextInputType.number,
                  validators: [
                    Validators.required(),
                    Validators.validLength(5,
                        errorText: "Phone number should be of length 5"),
                  ],
                  onFieldSubmitted: (_) {
                    fieldFocusChange(
                        context, _phoneNumberFocusNode, _emailAddressFocusNode);
                  },
                  onSaved: (value) {
                    signUpDetails.phoneNumber = value;
                  },
                ),
              ),
              SizedBox(height: 20),
              Container(
                width: 250,
                child: AutoValidatingTextFormField(
                  decoration: InputDecoration(
                    hintText: "Email Addres",
                  ),
                  focusNode: _emailAddressFocusNode,
                  controller: _emailAddressController,
                  keyboardType: TextInputType.emailAddress,
                  validators: [
                    Validators.required(),
                    Validators.pattern(
                        "^[a-zA-Z0-9.!#\$%&'*+/=?^_`{|}~-]+@[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,253}[a-zA-Z0-9])?(?:\.[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,253}[a-zA-Z0-9])?)*\$",
                        errorText: "Invalid email address."),
                  ],
                  onSaved: (value) {
                    signUpDetails.emailAddress = value;
                  },
                  onFieldSubmitted: (_) {
                    fieldFocusChange(
                        context, _emailAddressFocusNode, _passwordFocusNode);
                  },
                ),
              ),
              SizedBox(height: 20),
              Container(
                width: 250,
                child: AutoValidatingTextFormField(
                  decoration: InputDecoration(
                    hintText: "Password",
                  ),
                  obscureText: true,
                  focusNode: _passwordFocusNode,
                  controller: _passwordController,
                  keyboardType: TextInputType.text,
                  validators: [
                    Validators.required(),
                  ],
                  onSaved: (value) {
                    signUpDetails.password = value;
                  },
                  onFieldSubmitted: (_) {
                    fieldFocusChange(
                        context, _passwordFocusNode, _confirmPasswordFocusNode);
                  },
                ),
              ),
              SizedBox(height: 20),
              Container(
                width: 250,
                child: AutoValidatingTextFormField(
                  decoration: InputDecoration(
                    hintText: "Confirm Password",
                  ),
                  keyboardType: TextInputType.text,
                  focusNode: _confirmPasswordFocusNode,
                  controller: _confirmPasswordController,
                  obscureText: true,
                  validators: [
                    Validators.required(),
                    Validators.matches(
                      _passwordController,
                    )
                  ],
                ),
              ),
              SizedBox(height: 40),
              RaisedButton(
                child: Text(
                  "Sign Up",
                ),
                onPressed: () {
                  if (_formKey.currentState.validate()) {
                    _scaffoldKey.currentState.showSnackBar(
                      SnackBar(
                        content: Text(
                          "Signing up",
                        ),
                        duration: const Duration(seconds: 1),
                      ),
                    );
                  }
                },
              ),
            ],
          ),
        ),
      ),
    );
  }

  @override
  void dispose() {
    super.dispose();
  }
}

void fieldFocusChange(
    BuildContext context, FocusNode currentFocus, FocusNode nextFocus) {
  currentFocus.unfocus();
  FocusScope.of(context).requestFocus(nextFocus);
}
